File: src/gather_data/NaverTopKeywords.py
import time

import datetime as dt 

from config.SeleniumSettings import SeleniumSettings 

class NaverTopKeywords:
    def __init__(self, selenium_object: SeleniumSettings) -> None: 
        self.selenium_object = selenium_object 
        self.selenium_object.driver_settings()

    def naver_top_keyword_settings_method(self, naver_search_url: str, keywords_list: list[str], keyword_filter_list: list[str]) -> None:
        self.naver_search_url    = naver_search_url 
        self.keywords_list       = keywords_list 
        self.keyword_filter_list = keyword_filter_list

        self.naver_top_data_status_check = {
            "date"               : [],
            "search_keyword"     : [],
            "completion_status"  : []
        }

        self.naver_top_keywords_dictionary = {
            "date"              : [],
            "search_keyword"    : [],
            "trending_keywords" : [],
            "term_ranking"      : []
        }

    def get_keyword_data(self) -> None:
        def xpath_switch() -> list[str]: 
            keyword_list = ["extraction_failed"]
            keyword_list = self.selenium_object.search_for_element("fds-ugc-body-popular-topic-scroller", "class").text.split("\n") if (self.selenium_object.check_for_element('fds-ugc-body-popular-topic-scroller', "class")) else keyword_list

            return keyword_list
        
        def inner_function(keyword: str, keywords_list: list[str]) -> None:
            keywords_list = [i for i in keywords_list if (i not in self.keyword_filter_list)]

            for (rank_num, sub_keyword) in enumerate(keywords_list):
                self.naver_top_keywords_dictionary["date"].append(str(dt.datetime.now())[0:19])
                self.naver_top_keywords_dictionary["search_keyword"].append(keyword)
                self.naver_top_keywords_dictionary["trending_keywords"].append(sub_keyword)
                self.naver_top_keywords_dictionary["term_ranking"].append(str(rank_num + 1).zfill(2))
        
        def status_check(keyword: str, keywords_list: list[str]) -> int:
            extraction_status = 1 if (len(keywords_list) > 1) else 0
            
            self.naver_top_data_status_check["date"].append(str(dt.datetime.now())[0:19])
            self.naver_top_data_status_check["search_keyword"].append(keyword)
            self.naver_top_data_status_check["completion_status"].append(extraction_status)

            return extraction_status
                
        for keyword in self.keywords_list:
            self.selenium_object.driver.get(self.naver_search_url.format(keyword))
            time.sleep(5)
            keywords_list     = xpath_switch()
            extraction_status = status_check(keyword, keywords_list)
            inner_function(keyword, (lambda extraction_status, keywords_list: keywords_list if (extraction_status == 1) else ["extraction_failed"])(extraction_status, keywords_list))

File: java/gatherData/NaverTopKeywords.java
package org.gatherData;

import java.lang.Thread;
import java.time.Duration;
import java.time.LocalDateTime;
import java.util.Arrays;
import org.json.JSONArray;
import org.json.JSONObject;
import org.config.SeleniumSettings;

public class NaverTopKeywords {
    String   naverSearchUrl;
    String[] keywordsList, keywordFilterList;

    public JSONArray        naverTopDataStatusArray, naverTopKeywordsDictionary;
    public SeleniumSettings seleniumObject;

    NaverTopKeywords(SeleniumSettings seleniumObject) {
        this.seleniumObject = seleniumObject;
        seleniumObject.driverSettings();
    }

    public void naverTopKeywordsSettingsMethod(String naverSearchUrl, String[] keywordsList, String[] keywordFilterList) {
        this.naverSearchUrl             = naverSearchUrl;
        this.keywordsList               = keywordsList;
        this.keywordFilterList          = keywordFilterList;
        this.naverTopDataStatusArray    = new JSONArray();
        this.naverTopKeywordsDictionary = new JSONArray();
    }

    String[] xpathSwitch() throws Exception {
        String[] keywordList = new String[] {"extraction_failed"};

        return (seleniumObject.checkForElement("fds-ugc-body-popular-topic-scroller", "class")) ? seleniumObject.searchForElement("fds-ugc-body-popular-topic-scroller", "class").getText().split("\n") : keywordList;
    }

    int statusCheck(String keyword, String[] keywordsList) {
        JSONObject singleInnerJson  = new JSONObject();
        Integer    extractionStatus = (keywordsList.length > 1) ? 1 : 0;

        singleInnerJson.put("date",              String.valueOf(LocalDateTime.now()).substring(0, 10));
        singleInnerJson.put("search_keyword",    keyword);
        singleInnerJson.put("completion_status", extractionStatus);

        return extractionStatus;
    }

    void innerFunction(String keyword, String[] keywordsList) {
        int counterVar = 1;

        for (String subKeyword : keywordsList) {
            if (!Arrays.asList(keywordFilterList).contains(subKeyword)) {
                JSONObject naverTopDataJObject = new JSONObject();
                naverTopDataJObject.put("date", String.valueOf(LocalDateTime.now()).substring(0, 10));
                naverTopDataJObject.put("search_keyword", keyword);
                naverTopDataJObject.put("trending_keywords", subKeyword);
                naverTopDataJObject.put("term_ranking", String.format("%02d", counterVar));
                naverTopKeywordsDictionary.put(naverTopDataJObject);
                counterVar += 1;
            }
        }
    }

    public void getKeywordData() throws InterruptedException, Exception {
        for (String keyword : keywordsList) {
            seleniumObject.driver.get(String.format(naverSearchUrl, keyword));
            Thread.sleep(Duration.ofSeconds(5));
            String[] keywordsList      = xpathSwitch();
            Integer  extractionStatus  = statusCheck(keyword, keywordsList);
            String[] finalizedKeywords = (extractionStatus == 1) ? keywordsList : new String[] {"extraction_failed"};
            innerFunction(keyword, finalizedKeywords);
        }
    }
}
